packages/server/modules/shared/authz.ts

```diff
@@ -0,0 +1,293 @@
+import { Scopes } from '@/modules/core/helpers/mainConstants'
+import { getRoles } from '@/modules/shared'
+import { getStream } from '@/modules/core/services/streams'
+
+import {
+  BaseError,
+  ForbiddenError as FE,
+  UnauthorizedError as UE,
+  ContextError as CE,
+  BadRequestError as BRE
+} from '@/modules/shared/errors'
+// import { getbAllRoles } from '../core/services/generic'
+
+interface AuthResult {
+  authorized: boolean
+}
+
+interface AuthFailedResult extends AuthResult {
+  authorized: false
+  error: BaseError | null
+  fatal?: boolean
+}
+
+interface Stream {
+  role?: StreamRoles
```

Comment:
Better to keep DB entity types in some centralized place so that they can be re-used everywhere (and they def will be).

I've kept them (TS types or JSDoc Typedefs) in these places before:
- In the entity's repository module (e.g. `ServerInviteRecord` in `modules/serverinvites/repositories/index.js`)
- In the relevant speckle module's types helper (in this case - `@/modules/core/helpers/type.ts`

Additionally - you shouldn't mark fields in these kinds of types as optional (with ?), cause the DB schema being a strictly defined relational schema - all of the fields will always be there. Sure they might be null, then you can mark the value as Nullable<T>, but they can't be optional or undefined.

You might say that you might have a more limited query that only pulls specific fields from the DB, but in that case - wouldn't all of the fields be optional? Any and all fields can be omitted from a query, so theoretically they'd all have to be optional then.

The way I suggest dealing with this is making sure that this interface defines all of the fields with none of them being optional and with values correctly marked as Nullable, when it's possible, and then if you ever need to have a more limited version of this time you can use the `Pick`, `Omit` or `Partial` TS utility types to specify that some of the fields don't have to be there.

Reply:
yeah, that is true, but since duck typing I thought its fine. I sort of like the strictly necessary interface definition.
Plus this doesn't match the db record. The role? is there, because that is dynamically patched to the db record by the `getStream` function from `@/modules/core/repositories/streams.js`

Anyhow, we can start to define these types somewhere, my reasoning here was, to define the least amount needed for the auth context

Reply:
BTW In my own branch I've already created a `core` types helper at `@/modules/core/helpers/types.ts`, and I already defined a `StreamRecord` type there, so I guess the best path forward would be for you to either do the same and solve the merge conflict afterwards, or leave your Stream type as local to your module

Reply:
> yeah, that is true, but since duck typing I thought its fine. I sort of like the strictly necessary interface definition. Plus this doesn't match the db record. The role? is there, because that is dynamically patched to the db record by the `getStream` function from `@/modules/core/repositories/streams.js`
> 
> Anyhow, we can start to define these types somewhere, my reasoning here was, to define the least amount needed for the auth context

sure, if it's something very specific to your own module and not really reusable, then you can keep it there. considering the other comment I just made, it might be the best option